update the kubernetes tls documentation for use with the helm chart (#679)
redpanda-data/documentation: docs/security/kubernetes-tls.mdx

---
title: Configuring TLS on Kubernetes
---

<head>
    <meta name="title" content="Configuring TLS on Kubernetes | Redpanda Docs"/>
    <meta name="description" content="Encrypt with Transport Layer Security (TLS) on Kubernetes."/>
</head>

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

> Redpanda supports Transport Layer Security (TLS) encryption on Kubernetes. For certificate management, the Redpanda Helm chart uses cert-manager with either a self-signed Issuer or your own custom Issuer.

Author: Proper noun. Issuer is the name of the cert-manager resource being configured.

:::note

> This page uses the latest recommended Helm chart for configuring security. For information about using the older Redpanda Operator, see [Redpanda Operator](../../reference/redpanda-operator).

Reviewer: ```suggestion
This page uses the recommended Helm chart for configuring security. For information about using older versions of Redpanda Operator, see [Redpanda Operator](../../reference/redpanda-operator).
```

Author: It's not older versions though. It's an entirely different thing. I agree with removing "latest".

Reviewer 2: fixed with Joe's suggestion

:::

## Use self-signed Issuer

> By default, the Redpanda Helm chart uses cert-manager to create self-signed certificates. To enable TLS, simply enable it when installing or upgrading Redpanda with the Helm chart.

Reviewer: ```suggestion
By default, the Redpanda Helm chart uses cert-manager to create self-signed certificates. To enable TLS, enable it when installing or upgrading Redpanda with the Helm chart.
```

See [GitHub](https://github.com/redpanda-data/helm-charts/blob/main/redpanda/values.yaml) for the default values.

### Prerequisites

Install cert-manager:

```bash
helm repo add jetstack https://charts.jetstack.io
helm repo update
helm install cert-manager jetstack/cert-manager \
    --set installCRDs=true \
    --namespace cert-manager \
    --create-namespace
```

### Enable TLS

You can enable TLS in the Redpanda Helm chart using either command line flags or a YAML file to override default values.

<Tabs>
  <TabItem value="flags" label="CLI flags" default>

During install or upgrade, enable TLS configuration:

```bash
helm upgrade --install redpanda redpanda/redpanda -n redpanda --create-namespace \
  --set tls.enable=true --set external.domain=mydomain.dom
```

  </TabItem>
  <TabItem value="values" label="YAML values">

Create a YAML file containing the values to override from the defaults.

`tls_enable.yaml`

```yaml
tls:
  enable: true
external:
  domain: mydomain.dom
```

> During install or upgrade, reference that TLS configuration values file:

Reviewer: ```suggestion
During install or upgrade, reference the TLS configuration values file:
```

```bash
helm upgrade --install redpanda redpanda/redpanda -n redpanda --create-namespace \
  --values tls_enable.yaml
```

:::tip

> The values flag can be repeated, allowing you to keep individual value overrides logically separated by file.

Reviewer: ```suggestion
You can repeat the values flag, allowing you to keep individual value overrides logically separated by file.
```

:::

  </TabItem>
</Tabs>

### Validate connection

Retrieve the root certificate authority (CA) to use with clients:

```bash
kubectl -n redpanda get secret redpanda-default-root-certificate -o go-template='{{ index .data "ca.crt" | base64decode }}' > ca.crt
```

Use that root CA to validate your connection to Redpanda:

```bash
rpk topic list --tls-enabled --tls-truststore=ca.crt ...
```

## Use custom Issuer

You can use your own Issuer or ClusterIssuer to take advantage of an ACME certificate provider like Let's Encrypt or to use an intermediate CA.

### Prerequisites

> You must have a cert-manager Issuer in the Redpanda namespace or ClusterIssuer in your cluster. 

Reviewer: I looked at other sites, and most use "issuer"--not sure why "Issuer" is being capitalized in this topic file?

Author: Because it's a proper noun. It's the resource like Pod, Service, or Deployment.

See the [cert-manager documentation](https://cert-manager.io/docs/concepts/issuer/) for details. 

### Enable TLS

Create a YAML file containing the values to override the defaults.

`tls_enable.yaml`

```yaml
tls:
  enable: true
  certs:
    default:
      issuerRef:
        name: my-custom-issuer
        kind: ClusterIssuer
external:
  domain: mydomain.dom
```

> During install or upgrade, reference that TLS configuration values file:

Reviewer: ```suggestion
During install or upgrade, reference the TLS configuration values file:
```

  </TabItem>
  <TabItem value="values" label="YAML values">

> Create a YAML file containing the values to override from the defaults. For example:

Author: We don't want "For example:", we want them to follow these directions explicitly. If they treat it as an example and choose a different filename, the instruction we give them below will fail.
```suggestion
Create a YAML file containing the values to override from the defaults.
```

rpk topic list --tls-enabled --tls-truststore=ca.crt ...
```

## Use custom Issuer

You can use your own Issuer or ClusterIssuer to take advantage of an ACME certificate provider like Let's Encrypt or to use an intermediate CA.

### Prerequisites

You must have a cert-manager Issuer in the Redpanda namespace or ClusterIssuer in your cluster. 

See the [cert-manager documentation](https://cert-manager.io/docs/concepts/issuer/) for details. 

### Enable TLS

> Create a YAML file containing the values to override the defaults. For example:

Author: We don't want "For example:", we want them to follow these directions explicitly. If they treat it as an example and choose a different filename, the instruction we give them below will fail.
```suggestion
Create a YAML file containing the values to override the defaults.
```
